ref(replay): Make Session a POJO

mydea opened the pull request:
While the replay session was made a class here: https://github.com/getsentry/sentry-replay/pull/91, since we landed https://github.com/getsentry/sentry-javascript/pull/6315, it's not really necessary anymore to have this be a class.

This PR refactors this to a simple POJO with a function to easily generate it with defaults etc. 
Also, some things like the `toJSON()` where only used in tests. We should generally avoid this, and instead provide some test utils specifically for these things (where necessary), to avoid shipping this to our end users.

Lms24 reviewed (approved):
Nice!

Lms24 commented:
Eventually, we'd like to get rid of nullish coalescence operators to save some bundle size but we can tackle this when we get to bundle size optimization later on as well.

mydea commented:
> Eventually, we'd like to get rid of nullish coalescencing operators to save some bundle size but we can tackle this when we get to bundle size optimization later on as well.

Good point, I'll implement that as well while I'm at it! Check it out.

AbhiPrasad commented on packages/replay/src/session/Session.ts:
```diff
@@ -24,59 +24,41 @@ interface SessionObject {
+/**
+ * Get a session with defaults & applied sampling.
+ */
+export function makeSession(session: Partial<Session>, { sessionSampleRate, errorSampleRate }: SampleRates): Session {
```
nit: If we get rid of the obj arg in favour of direct args with `sessionSampleRate` and `errorSampleRate` we'll save some bytes because we can minify those variables.

This makes sense to do if `makeSession` is not user facing, as then we can change the method signature in the future.

AbhiPrasad reviewed (approved):
Awesome! Feels like v7 work all over again, we right back to it!

mydea commented:
So I adjusted this a bit based on feedback from @AbhiPrasad . basically, I extracted the sampling decision out into a dedicated function, that only needs to be called when creating a new session. No need to actually call this in `fetchSession`, as there the session has to have `sampled` already. This allowed to slightly reduce bundle size.

Lms24 commented on packages/replay/src/session/Session.ts:
```diff
@@ -24,59 +23,48 @@ interface SessionObject {
+export function makeSession(session: Partial<Session> & { sampled: Sampled }): Session {
+  const now = new Date().getTime();
+  const id = session.id || uuid4();
+  const started = session.started || now;
```
Since tests are passing and I think for this case there is no difference, I guess the risk is low but just to confirm:

Is this change from the nullish coalescing operator safe here? `||` has different behaviour for falsy values than `??`. (but as I said, probably it's fine)

AbhiPrasad replied:
I guess this behavior is technically not correct if `started` is 0, but I think that's an edge case we don't worry about.

Lms24 replied:
Exactly my thoughts

mydea replied:
Yes, technically it is not the same, but I think we don't care that you cannot set a time of January 1st 1979. There is actually a test that I adjusted for this. I will add a comment to this file as well explaining that this is a conscious decision!

Lms24 reviewed (approved):
LGTM. I like the `sampled` change

AbhiPrasad commented on packages/replay/MIGRATION.md:
```diff
@@ -52,3 +52,8 @@ import something from '@sentry/replay/submodule';
+
+## Session object is now a POJO (https://github.com/getsentry/sentry-javascript/pull/6417)
+
+The `Session` object exported from Replay is now a POJO, instead of a class.
```
also a nit - I would just say a plain object, might be easier for developers to understand (especially those who are more unfamiliar with JS).

AbhiPrasad commented on packages/replay/src/session/Session.ts:
```diff
@@ -24,59 +23,48 @@ interface SessionObject {
+/**
+ * Get the sampled status for a session based on sample rates & current sampled status.
+ */
+export function sampleSession(
```
l: `getSessionSampleType`

AbhiPrasad commented on packages/replay/src/session/Session.ts:
```diff
@@ -24,59 +23,48 @@ interface SessionObject {
+ * Get the sampled status for a session based on sample rates & current sampled status.
+ */
+export function sampleSession(
+  sampled: Sampled | undefined,
```
m: now with this refactor, isn't the only usage of `sampleSession` when it's undefined?

Does this mean we can even remove this arg and the early return?

mydea replied:
hmm, good point! Def. makes sense 👍
